File: btypes_primitives/src/main/python_pyrsistant_immutable/btypes/BString.py
from btypes.BBoolean import *

class BString:

	def __init__(self, value):
		self.__value = value

	def getValue(self):
		return value

	def __eq__(self, other: 'BString') -> 'BBoolean':
		return BBoolean(self.__value == other.__value)

	def equals(self, other) -> 'bool':
		return self.__eq__(other)

	def length(self) -> 'int':
		return len(self.__value)

	def isEmpty(self) -> 'bool':
		return not self.__value

	def hashCode(self):
		hash(self.__value)

    def equal(self, other: 'BString') -> 'BBoolean':
        return BBoolean(self.__value == other.__value)

    def unequal(self, other: 'BString') -> 'BBoolean':
        return BBoolean(self.__value != other.__value)

	def toString(self) -> 'str':
		return '"' + self.__value + '"'

	def isCase(self, other) -> 'bool':
		return self.__value.equals(other)

	def isString(self) -> 'BBoolean':
		return BBoolean(True)

	def isNotString(self) -> 'BBoolean':
		return BBoolean(False)

File: visualizations/LightModel/btypes/BString.js
import { BBoolean } from "https://favu100.github.io/b2program/visualizations/LightModel/btypes/BBoolean.js";
export class BString {
    constructor(value) {
        this.value = value;
    }
    getValue() {
        return this.value;
    }
    equals(o) {
        if (o == null || o instanceof BString) {
            return false;
        }
        return this.value === o.value;
    }
    equal(o) {
        return new BBoolean(this.equals(o));
    }
    unequal(o) {
        return new BBoolean(!this.equals(o));
    }
    length() {
        return this.value.length;
    }
    isEmpty() {
        return this.value === "";
    }
    toString() {
        return '"' + this.value + '"';
    }
    isCase(o) {
        return this.value == o;
    }
    isString() {
        return new BBoolean(true);
    }
    isNotString() {
        return new BBoolean(false);
    }
    hashCode() {
        let hash = 0;
        if (this.value.length === 0) {
            return hash;
        }
        for (let i = 0; i < this.value.length; i++) {
            let chr = this.value.charCodeAt(i);
            hash = ((hash << 5) - hash) + chr;
            hash |= 0; // Convert to 32bit integer
        }
        return hash;
    }
    ;
}
